src/pycodehash/hashing.py

```diff
@@ -130,32 +133,63 @@ def hash_location(self, location: Location, project: Project) -> str:
         self.func_store[location] = function_hash
         return function_hash
 
-    def _get_location_and_project(self, func: Callable) -> tuple[Location, Project]:
-        project = self.project_store.get_from_func(func)
+    def _get_location_and_project(self, func: FunctionType) -> tuple[Location, Project]:
+        """Hash a Python function
+
+        Args:
+            func: the Python function
+
+        Raises:
+            TypeError: when `func` is a `BuiltinFunctionType` as these do not have accessible source code
+            ValueError: when the source code for `func` cannot be found but it is not a `BuiltinFunctionType`
+
+        Returns:
+            location, project: the location of the function definition and the project it belongs to
+
+        """
+        # exit path for when we cannot hash the source
+        if isinstance(func, BuiltinFunctionType):
+            fname = self._get_func_name(func)
+            raise TypeError(f"builtin function `{fname}` cannot be hashed as there is no Python source code.")
```

Comment:
Python includes the raise in the default traceback (and formatters like Rich and IPython do too).
Passing a string to the constructor duplicates the traceback, which can make the traceback less readable.

Simply assign the string to a variable before constructing the exception is one way to fix this:
```
msg = f"builtin function `{fname}` cannot be hashed as there is no Python source code."
raise TypeError(msg)
```
Alternatively, we could define custom exceptions. For this TypeError I do think that is overkill.

Before:
```python
Traceback (most recent call last):
  File "tmp.py", line 2, in <module>
    raise TypeError(f"builtin function `{fname}` cannot be hashed as there is no Python source code.")
TypeError: builtin function `hello` cannot be hashed as there is no Python source code.
```

After:
```python
Traceback (most recent call last):
  File "tmp.py", line 2, in <module>
    raise TypeError(msg)
TypeError: builtin function `hello` cannot be hashed as there is no Python source code.
```